### PR TITLE
Wrap basic types in a struct

### PR DESCRIPTION
**Pull request by liamsi**
ref #204

Note: Instead of switching over all possible cases via reflection (and maybe recursively), this simply adds the bytes. This PR just emulates a wrapping struct.

### CODE REVIEW

File: amino.go [hunk 1]
```diff
@@ -203,21 +211,43 @@ func (cdc *Codec) MarshalBinaryBare(o interface{}) ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
-	err = cdc.encodeReflectBinary(buf, info, rv, FieldOptions{BinFieldNum: 1}, true)
-	if err != nil {
-		return nil, err
+	// in the case of of a repeated struct (e.g. type Alias []SomeStruct),
+	// we do not need to prepend with `(field_number << 3) | wire_type` as this
+	// would need to be done for each struct and not only for
```
Comment (liamsi) [inline amino.go]:
```suggestion
	// would need to be done for each struct and not only for the first.
```

File: amino.go [hunk 2]
```diff
@@ -387,10 +444,6 @@ func (cdc *Codec) MarshalJSON(o interface{}) ([]byte, error) {
 
 	// disfix wrapper continued...
 	if info.Registered {
-		// Part 2:
-		if err != nil {
-			return nil, err
-		}
```
Comment (liamsi) [inline amino.go:393]: Not sure what this was supposed to do ... The err would always have been nil here.

File: json_test.go [hunk 3]
```diff
@@ -172,24 +172,8 @@ type innerFP struct {
 }
 
 func TestUnmarshalMap(t *testing.T) {
-	binBytes := []byte(`dontcare`)
 	obj := new(map[string]int)
 	cdc := amino.NewCodec()
-	// Binary doesn't support decoding to a map...
-	// TODO: move out binary tests from json_test.go ...
-	assert.Panics(t, func() {
```
Comment (liamsi) [inline json_test.go:180]: These were not deleted but moved into binary_test.go

File: amino.go [hunk 4]
```diff
@@ -2,21 +2,29 @@ package amino
...
+
+	// we use this time to init. a zero value (opposed to reflect.Zero which gives time.Time{} / 01-01-01 00:00:00)
+	zeroTime time.Time
 
-// we use this time to init. a zero value (opposed to reflect.Zero which gives time.Time{} / 01-01-01 00:00:00)
-var zeroTime time.Time
+	// NotPointerErr is thrown when you call a method that expects a pointer, e.g. Unmarshal
+	NotPointerErr = errors.New("expected a pointer")
+
+	UnknownBasicTypeErr = fmt.Errorf("got unknown type: ")
```
Comment (jordaaash) [inline amino.go]: Is this error used anywhere?

Screen Shot 2019-05-22 at 12 41 02 PM

File: amino.go [hunk 5]
```diff
@@ -2,21 +2,29 @@ package amino
...
+	// Global methods for global sealed codec.
+	gcdc *Codec
+
+	// we use this time to init. a zero value (opposed to reflect.Zero which gives time.Time{} / 01-01-01 00:00:00)
+	zeroTime time.Time
 
-// we use this time to init. a zero value (opposed to reflect.Zero which gives time.Time{} / 01-01-01 00:00:00)
-var zeroTime time.Time
+	// NotPointerErr is thrown when you call a method that expects a pointer, e.g. Unmarshal
+	NotPointerErr = errors.New("expected a pointer")
```
Comment (jordaaash) [inline amino.go:24]: I'm not familiar with stdlib's `errors` vs. `pkg/errors` vs. `fmt.Errorf`, but I saw this dependency isn't used elsewhere. Worth adding a new library for?

File: amino.go (hunk 4 above)
Comment (liamsi) [reply amino.go]: Good catch! That's a relict from another approach (actually wrapping each basic type in a struct instead of just adding the necessary bytes). Removed.

File: amino.go (hunk 5 above)
Comment (liamsi) [reply amino.go:24]: pkg/errors allows you to wrap the error with some context instead of just passing it on. This helps callers to understand exactly where the error occurred. Although you can achieve sth. similar with fmt.Errorf, I think error.Wrap is slightly more convenient (and it also adds the call stack if you want to).

File: amino.go (hunk 4 above)
Comment (Liamsi) [reply amino.go]: This is an excellent read on errors and touches the different error packages too: https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully

File: amino.go (hunk 5 above)
Comment (liamsi) [reply amino.go:24]: This is an excellent read on errors and touches the different error packages too: https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully

File: amino.go [hunk 6]
```diff
@@ -340,6 +371,28 @@ func (cdc *Codec) UnmarshalBinaryBare(bz []byte, ptr interface{}) error {
...
+	isRepeatedStruct := info.Type.Kind() == reflect.Slice && info.Type.Elem().Kind() == reflect.Struct
+	isPointerToStruct := isPtr && rv.Elem().Kind() != reflect.Struct
+	if rv.Kind() != reflect.Struct && !isRepeatedStruct && len(bz) > 0 && (rv.Kind() != reflect.Interface) && !isPointerToStruct {
+
+		fnum, typ, n, err := decodeFieldNumberAndTyp3(bz)
+		if err != nil {
+			return errors.Wrap(err, "could not decode field number and type")
+		}
+		if fnum != 1 {
+			fmt.Println("What is it?", rv.Kind())
```
Comment (liamsi) [inline amino.go]: Forgot to delete this line.

File: binary-encode.go [hunk 7]
```diff
@@ -413,27 +413,12 @@ func (cdc *Codec) encodeReflectBinaryStruct(w io.Writer, info *TypeInfo, rv refl
...
-					return
-				}
-				lAfterValue := buf.Len()
-
-				if !frvIsPtr && !fopts.WriteEmpty && lBeforeValue == lAfterValue-1 && buf.Bytes()[buf.Len()-1] == 0x00 {
-					// rollback typ3/fieldnum and last byte if
-					// not a pointer and empty:
-					buf.Truncate(lBeforeKey)
-				}
-
```
Comment (liamsi) [inline binary-encode.go:436]: This logic was moved to `writeFieldIfNotEmpty` as it is now used in two places (when wrapping a basic type as a field of a struct).

**Review by alexanderbez — APPROVED** [review]
ACK on first passthrough.

File: amino.go [hunk 8]
```diff
@@ -340,6 +371,26 @@ func (cdc *Codec) UnmarshalBinaryBare(bz []byte, ptr interface{}) error {
 		}
 		bz = bz[4:]
 	}
+	isRepeatedStruct := info.Type.Kind() == reflect.Slice && info.Type.Elem().Kind() == reflect.Struct
```
Comment (alexanderbez) [inline amino.go]: Maybe we ought to have an auxiliary function given an `info` that returns a bool if it's a repeated struct.

Comment (liamsi) [reply amino.go]: Changed. Also, added another test case and a few related changes in 8764911

**Comment by liamsi** [comment]
I just found another edge-case where this is not complete: when you have an interface and the concrete implementation is a type alias. Will add a test-case and fix shortly.

**Comment by alexanderbez** [comment]
What's the best way to fully test this for all corner cases @Liamsi ?

**Comment by liamsi** [comment]
More extensive & more coverage via unit-tests (and maybe extend the fuzzing code too). This would also have been caught, if one for instance ran tendermint / gaia using the modified version in this PR (b/c pub-keys are interfaces and their concrete implementations are type aliases).

I think, we can merge this PR as is. I'll do the change mentioned above in a follow-up PR ([issue](https://github.com/tendermint/go-amino/issues/273)). Before, we do a release, it would be great if we ran and tested it with tendermint and gaia before (additionally to the fuzzer and more unit tests).

**Comment by alexanderbez** [comment]
Not a problem, I can easily test that out @Liamsi
